lib/WTSI/NPG/HTS/Samtools.pm

```diff
@@ -0,0 +1,195 @@
+
+=cut
+
+sub put_xam {
+    my ($local_path, $remote_path, $reference_path) = @_;
+
+    my $tmpdir = tempdir('put_xam_XXXXXX', TMPDIR => 1, CLEANUP => 1);
+    my ($name, $path, $suffix) = fileparse($remote_path, qr/[.][^.]*$/msxi);
+
+    my @samtools_args = ('view', '-T', $reference_path);
+    $suffix = fc $suffix;
+
+    if ($suffix eq '.sam') {
+        push @samtools_args, '-h' # Output SAM, retain the header
+    }
+    elsif ($suffix eq '.bam') {
+        push @samtools_args, '-b' # Output BAM
+    }
+    elsif ($suffix eq '.cram') {
+        push @samtools_args, '-C' # Output CRAM
+    }
+    else {
+        croak "Unsupported file type: '$suffix'";
+    }
+
+    push @samtools_args, '-o', "$tmpdir/$name", $local_path;
+
+    # Runnable handles logging and error checking
+    # First run samtools to create a local temp file
+    WTSI::DNAP::Utilities::Runnable->new(
+        arguments  => \@samtools_args,
+        executable => $DEFAULT_SAMTOOLS_EXECUTABLE)->run;
+
+    # Then run iput to put the temp file into iRODS
+    WTSI::DNAP::Utilities::Runnable->new(
+        arguments  => ['-f', '-k', '-a', "$tmpdir/$name", $remote_path],
+        executable => $DEFAULT_IPUT_EXECUTABLE)->run;
+
+    return;
+}
+
+=head2 get_xam_header
+
+  Arg [1]    : Str $remote_path
+               The path to the remote SAM/BAM/CRAM file in iRODS.
+
+  Example    : my $header = get_xam_header('/path/to/file.bam');
+
+  Description: Retrieves the header of a SAM/BAM/CRAM file stored in iRODS.
+               Uses `iget` to fetch the file and `samtools` to extract the header.
+
+  Returntype : ArrayRef[Str]
+               An array reference containing the header lines as strings.
+
+  Throws     : Croaks if the `iget` or `samtools` command fails.
+
+=cut
+
+sub get_xam_header {
+    my ($remote_path) = @_;
+
+    my $cmd = "$DEFAULT_IGET_EXECUTABLE $remote_path - | $DEFAULT_SAMTOOLS_EXECUTABLE head";
+    my $exit;
+    my ($stdout, $stderr) = capture {
+        $exit = system $cmd
+    };
+
+    if ($exit != 0) {
+        croak "Failed to run '$cmd': $ERRNO : $stderr";
+    }
+
+    return [split /\n/msx, $stdout];
+}
+
+
+=head2 get_xam_records
+
+  Arg [1]    : Str $remote_path
+               The path to the remote SAM/BAM/CRAM file in iRODS.
+
+  Arg [2]    : Int $num_records (optional)
+               The number of records to retrieve. Defaults to 1024 if not provided.
+
+  Example    : my $records = get_xam_records('/path/to/file.bam', 500);
+
+  Description: Retrieves a specified number of records from a SAM/BAM/CRAM file
+               stored in iRODS. Uses `iget` to fetch the file and `samtools` to
+               extract the records.
+
+  Returntype : ArrayRef[Str]
+               An array reference containing the retrieved records as strings.
+
+  Throws     : Croaks if the `iget` or `samtools` command fails.
+
+=cut
+
+sub get_xam_records {
+    my ($remote_path, $num_records) = @_;
+    $num_records //= 1024; # Default to 1024 records
```

Comment:
There is no pattern matching on this variable to check whether the string has a number in it. After calling this function with any string, this assignment will be successful and `samtools` may fail.
The pattern matching is checked only by the caller `_get_reads` in `AlnDataObject.pm`.
Is this the expected behaviour?